File: go/IntToRoman/IntToRoman.go
package main

import (
	"fmt"
	"strings"
)

// Time -> 0(N)
func intToRoman(num int) string {
	result := &strings.Builder{}
	for num > 0 {
		if num/1000 >= 1 {
			result.WriteString("M")
			num -= 1000
		} else if num/100 >= 1 {
			if num >= 400 && num <= 499 {
				result.WriteString("CD")
				num -= 400
			} else if num >= 900 && num <= 999 {
				result.WriteString("CM")
				num -= 900
			} else if num >= 500 {
				result.WriteString("D")
				num -= 500
			} else {
				result.WriteString("C")
				num -= 100
			}
		} else if num/10 >= 1 {
			if num >= 40 && num <= 49 {
				result.WriteString("XL")
				num -= 40
			} else if num >= 90 && num <= 99 {
				result.WriteString("XC")
				num -= 90
			} else if num >= 50 {
				result.WriteString("L")
				num -= 50
			} else {
				result.WriteString("X")
				num -= 10
			}
		} else {
			if num == 4 {
				result.WriteString("IV")
				num -= 4
			} else if num == 9 {
				result.WriteString("IX")
				num -= 9
			} else if num >= 5 {
				result.WriteString("V")
				num -= 5
			} else {
				result.WriteString("I")
				num -= 1
			}
		}
	}
	return result.String()
}

func main() {
	fmt.Println(intToRoman(3979))
}

File: python/intToRoman.py
def intToRoman(num: int) -> str:
    result = []
    while num > 0:
        if (num // 1000) >= 1:
            result.append("M")
            num -= 1000
        elif (num // 100) >= 1:
            if num >= 400 and num <= 499:
                result.append("CD")
                num -= 400
            elif num >= 900 and num <= 999:
                result.append("CM")
                num -= 900
            elif num >= 500:
                result.append("D")
                num -= 500
            else:
                result.append("C")
                num -= 100

        elif (num // 10) >= 1:
            if num >= 40 and num <= 49:
                result.append("XL")
                num -= 40
            elif num >= 90 and num <= 99:
                result.append("XC")
                num -= 90
            elif num >= 50:
                result.append("L")
                num -= 50
            else:
                result.append("X")
                num -= 10

        else:
            if num == 4:
                result.append("IV")
                num -= 4
            elif num == 9:
                result.append("IX")
                num -= 9
            elif num >= 5:
                result.append("V")
                num -= 5
            else:
                result.append("I")
                num -= 1

    return "".join(result)


print(intToRoman(3979))
